**Project Id : PTID-CDS-FEB-25-2469**

**Project Code : PRCP-1016-HeartDieseasePred**

**Project Name : HeartDieseasePrediction**

**Problem Statement**

Task 1:-Prepare a complete data analysis report on the given data.

Task 2:- Create a model predicting potential Heart Diseases in people using Machine Learning algorithms.

Task3:-Suggestions to the Hospital  to awake the predictions of heart diseases  prevent life threats.


 ### Summary of the project

The Heart Disease Prediction project leverages machine learning algorithms to predict the likelihood of heart disease in patients based on various health parameters. The model uses features such as age, gender, blood pressure, cholesterol levels, smoking habits, exercise-induced angina, and more. The primary goal is to aid early diagnosis and improve treatment decisions.

In [1]:
install xgboost

SyntaxError: invalid syntax (2373286279.py, line 1)

In [2]:
# Import important libraries
import pandas as pd
import numpy as np

# Import important libraries visualization
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

# Import libraries for handling categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

# Libraries to split data into Train and Test Data
from sklearn.model_selection import train_test_split

# Library for statistical operations
from scipy.stats import chi2_contingency, ttest_ind
from scipy import stats

# Import libraries to handle scaling
from sklearn.preprocessing import MinMaxScaler, StandardScaler
# Import models algorithms
from sklearn.model_selection import GridSearchCV, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.cluster import KMeans
from sklearn.svm import SVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import RandomForestRegressor, BaggingRegressor, AdaBoostRegressor, GradientBoostingRegressor
import xgboost as xgb
from xgboost import XGBRegressor

# Libraries to evaluate model
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

# Library to ignore warnings
import warnings
warnings.filterwarnings('ignore')

ModuleNotFoundError: No module named 'xgboost'

In [4]:
# Load the CSV files
labels_df = pd.read_csv('labels.csv')
values_df = pd.read_csv('values.csv')

# Merge the two DataFrames on the 'patient_id' column
data= pd.merge(values_df, labels_df, on='patient_id')

In [6]:
data

patient_id  slope_of_peak_exercise_st_segment               thal  \
0       0z64un                                  1             normal   
1       ryoo3j                                  2             normal   
2       yt1s1x                                  1             normal   
3       l2xjde                                  1  reversible_defect   
4       oyt4ek                                  3  reversible_defect   
..         ...                                ...                ...   
175     5qfar3                                  2  reversible_defect   
176     2s2b1f                                  2             normal   
177     nsd00i                                  2  reversible_defect   
178     0xw93k                                  1             normal   
179     2nx10r                                  1             normal   

     resting_blood_pressure  chest_pain_type  num_major_vessels  \
0                       128                2                  0   
1                       110                3                  0   
2                       125                4                  3   
3                       152                4                  0   
4                       178                1                  0   
..                      ...              ...                ...   
175                     125                4                  2   
176                     180                4                  0   
177                     125                3                  0   
178                     124                3                  2   
179                     160                3                  1   

     fasting_blood_sugar_gt_120_mg_per_dl  resting_ekg_results  \
0                                       0                    2   
1                                       0                    0   
2                                       0                    2   
3                                       0                    0   
4                                       0                    2   
..                                    ...                  ...   
175                                     1                    0   
176                                     0                    1   
177                                     0                    0   
178                                     1                    0   
179                                     0                    0   

     serum_cholesterol_mg_per_dl  oldpeak_eq_st_depression  sex  age  \
0                            308                       0.0    1   45   
1                            214                       1.6    0   54   
2                            304                       0.0    1   77   
3                            223                       0.0    1   40   
4                            270                       4.2    1   59   
..                           ...                       ...  ...  ...   
175                          254                       0.2    1   67   
176                          327                       3.4    0   55   
177                          309                       1.8    1   64   
178                          255                       0.0    1   48   
179                          201                       0.0    0   54   

     max_heart_rate_achieved  exercise_induced_angina  heart_disease_present  
0                        170                        0                      0  
1                        158                        0                      0  
2                        162                        1                      1  
3                        181                        0                      1  
4                        145                        0                      0  
..                       ...                      ...                    ...  
175                      163                        0                      1  
176                      117                

# Basic Checks

In [8]:
#It Display the first Five Rows
data.head()

patient_id  slope_of_peak_exercise_st_segment               thal  \
0     0z64un                                  1             normal   
1     ryoo3j                                  2             normal   
2     yt1s1x                                  1             normal   
3     l2xjde                                  1  reversible_defect   
4     oyt4ek                                  3  reversible_defect   

   resting_blood_pressure  chest_pain_type  num_major_vessels  \
0                     128                2                  0   
1                     110                3                  0   
2                     125                4                  3   
3                     152                4                  0   
4                     178                1                  0   

   fasting_blood_sugar_gt_120_mg_per_dl  resting_ekg_results  \
0                                     0                    2   
1                                     0                    0   
2                                     0                    2   
3                                     0                    0   
4                                     0                    2   

   serum_cholesterol_mg_per_dl  oldpeak_eq_st_depression  sex  age  \
0                          308                       0.0    1   45   
1                          214                       1.6    0   54   
2                          304                       0.0    1   77   
3                          223                       0.0    1   40   
4                          270                       4.2    1   59   

   max_heart_rate_achieved  exercise_induced_angina  heart_disease_present  
0                      170                        0                      0  
1                      158                        0                      0  
2                      162                        1                      1  
3                      181                        0                      1  
4                      145                        0                      0

In [10]:
#it display the last five rows 
data.tail()

patient_id  slope_of_peak_exercise_st_segment               thal  \
175     5qfar3                                  2  reversible_defect   
176     2s2b1f                                  2             normal   
177     nsd00i                                  2  reversible_defect   
178     0xw93k                                  1             normal   
179     2nx10r                                  1             normal   

     resting_blood_pressure  chest_pain_type  num_major_vessels  \
175                     125                4                  2   
176                     180                4                  0   
177                     125                3                  0   
178                     124                3                  2   
179                     160                3                  1   

     fasting_blood_sugar_gt_120_mg_per_dl  resting_ekg_results  \
175                                     1                    0   
176                                     0                    1   
177                                     0                    0   
178                                     1                    0   
179                                     0                    0   

     serum_cholesterol_mg_per_dl  oldpeak_eq_st_depression  sex  age  \
175                          254                       0.2    1   67   
176                          327                       3.4    0   55   
177                          309                       1.8    1   64   
178                          255                       0.0    1   48   
179                          201                       0.0    0   54   

     max_heart_rate_achieved  exercise_induced_angina  heart_disease_present  
175                      163                        0                      1  
176                      117                        1                      1  
177                      131                        1                      1  
178                      175                        0                      0  
179                      163                        0                      0

In [12]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 15 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   patient_id                            180 non-null    object 
 1   slope_of_peak_exercise_st_segment     180 non-null    int64  
 2   thal                                  180 non-null    object 
 3   resting_blood_pressure                180 non-null    int64  
 4   chest_pain_type                       180 non-null    int64  
 5   num_major_vessels                     180 non-null    int64  
 6   fasting_blood_sugar_gt_120_mg_per_dl  180 non-null    int64  
 7   resting_ekg_results                   180 non-null    int64  
 8   serum_cholesterol_mg_per_dl           180 non-null    int64  
 9   oldpeak_eq_st_depression              180 non-null    float64
 10  sex                                   180 non-null    int64  
 11  age                

In [15]:
data.describe()

slope_of_peak_exercise_st_segment  resting_blood_pressure  \
count                         180.000000              180.000000   
mean                            1.550000              131.311111   
std                             0.618838               17.010443   
min                             1.000000               94.000000   
25%                             1.000000              120.000000   
50%                             1.000000              130.000000   
75%                             2.000000              140.000000   
max                             3.000000              180.000000   

       chest_pain_type  num_major_vessels  \
count       180.000000         180.000000   
mean          3.155556           0.694444   
std           0.938454           0.969347   
min           1.000000           0.000000   
25%           3.000000           0.000000   
50%           3.000000           0.000000   
75%           4.000000           1.000000   
max           4.000000           3.000000   

       fasting_blood_sugar_gt_120_mg_per_dl  resting_ekg_results  \
count                            180.000000           180.000000   
mean                               0.161111             1.050000   
std                                0.368659             0.998742   
min                                0.000000             0.000000   
25%                                0.000000             0.000000   
50%                                0.000000             2.000000   
75%                                0.000000             2.000000   
max                                1.000000             2.000000   

       serum_cholesterol_mg_per_dl  oldpeak_eq_st_depression         sex  \
count                   180.000000                180.000000  180.000000   
mean                    249.211111                  1.010000    0.688889   
std                      52.717969                  1.121357    0.464239   
min                     126.000000                  0.000000    0.000000   
25%                     213.750000                  0.000000    0.000000   
50%                     245.500000                  0.800000    1.000000   
75%                     281.250000                  1.600000    1.000000   
max                     564.000000                  6.200000    1.000000   

              age  max_heart_rate_achieved  exercise_induced_angina  \
count  180.000000               180.000000               180.000000   
mean    54.811111               149.483333                 0.316667   
std      9.334737                22.063513                 0.466474   
min     29.000000                96.000000                 0.000000   
25%     48.000000               132.000000                 0.000000   
50%     55.000000               152.000000                 0.000000   
75%     62.000000               166.250000                 1.000000   
max     77.000000               202.000000                 1.000000   

       heart_disease_present  
count             180.000000  
mean                0.444444  
std                 0.498290  
min                 0.000000  
25%                 0.000000  
50%                 0.000000  
75%                 1.000000  
max                 1.000000

In [14]:
#it display the rows and column
data.shape

(180, 15)

In [16]:
#It is used to identify missing or null values in a dataset
data.isnull().sum()

patient_id                              0
slope_of_peak_exercise_st_segment       0
thal                                    0
resting_blood_pressure                  0
chest_pain_type                         0
num_major_vessels                       0
fasting_blood_sugar_gt_120_mg_per_dl    0
resting_ekg_results                     0
serum_cholesterol_mg_per_dl             0
oldpeak_eq_st_depression                0
sex                                     0
age                                     0
max_heart_rate_achieved                 0
exercise_induced_angina                 0
heart_disease_present                   0
dtype: int64

In [18]:
#Unique Values
data.apply(lambda x:len(x.unique()))

patient_id                              180
slope_of_peak_exercise_st_segment         3
thal                                      3
resting_blood_pressure                   36
chest_pain_type                           4
num_major_vessels                         4
fasting_blood_sugar_gt_120_mg_per_dl      2
resting_ekg_results                       3
serum_cholesterol_mg_per_dl             118
oldpeak_eq_st_depression                 34
sex                                       2
age                                      40
max_heart_rate_achieved                  73
exercise_induced_angina                   2
heart_disease_present                     2
dtype: int64

## EDA [Exploratory Data analysis]

### Univariate Analysis

In [20]:
import sweetviz as sv
my_report=sv.analyze(data)
my_report.show_html()

                                             |                                             | [  0%]   00:00 ->…

Report SWEETVIZ_REPORT.html was generated! NOTEBOOK/COLAB USERS: the web browser MAY not pop up, regardless, the report IS saved in your notebook/colab files.


## Insights on Univariate analysis

**Heart Disease Present**

- Binary classification, with 56% of the data indicating "No" (0) and 44% indicating "Yes" (1). This suggests a somewhat balanced target distribution, though slightly skewed toward "No".

**Fasting Blood Sugar > 120 mg/dl**

- 84% of patients have fasting blood sugar levels ≤ 120 mg/dl, while 16% have higher levels, which may indicate a potential risk factor for heart disease.

**Thalassemia Type (Thal)**

- Three categories: "Normal" (54%), "Reversible Defect" (41%), and "Fixed Defect" (4%). Most patients have a normal thalassemia type, with a significant portion showing reversible defects, which could be linked to heart disease.

**Max Heart Rate Achieved**

- The correlation with age is -0.39, suggesting that as age increases, the maximum heart rate tends to decrease, which is expected. There's also a weak correlation with cholesterol and resting blood pressure.

**Slope of Peak Exercise ST Segment**

- Moderate correlation with the target variable (heart disease presence), as well as with chest pain type. The slope is a critical feature in determining heart conditions during exercise.

**Chest Pain Type**
- Different types of chest pain have a strong correlation with heart disease presence, indicating its significance in predicting heart conditions.